Add support for gpsd API version 9 (gpsd 3.20)
https://bazaar.launchpad.net/~foxtrotgps-team/foxtrotgps/trunk/revision/320

Add support for gpsd API version 10 (gpsd 3.21)
https://bazaar.launchpad.net/~foxtrotgps-team/foxtrotgps/trunk/revision/327

gpsd API 9 has a different altitude reporting
https://bazaar.launchpad.net/~foxtrotgps-team/foxtrotgps/trunk/revision/328

deprecated STATUS_NO_FIX has been removed from.gpsd, based on
https://bazaar.launchpad.net/~foxtrotgps-team/foxtrotgps/trunk/revision/329

Index: src/gps_functions.c
--- src/gps_functions.c.orig
+++ src/gps_functions.c
@@ -753,12 +753,20 @@ cb_gpsd_data(GIOChannel *src, GIOCondition condition, 
 	{
 		gpsdata->satellites_used = libgps_gpsdata.satellites_used;
 		gpsdata->hdop = libgps_gpsdata.dop.hdop;
+#if GPSD_API_MAJOR_VERSION >= 9 /* API change. gpsd version 3.20 and subsequent. */
+		gpsdata->fix.time = libgps_gpsdata.fix.time.tv_sec + (libgps_gpsdata.fix.time.tv_nsec*1e-9);
+#else
 		gpsdata->fix.time = libgps_gpsdata.fix.time;
+#endif
 		if (isnan(gpsdata->fix.time))
 		{
 			gpsdata->fix.time = (time_t) 0;
 		}
+#if GPSD_API_MAJOR_VERSION >= 10
+		gpsdata->valid = (libgps_gpsdata.fix.mode >= MODE_2D);
+#else
 		gpsdata->valid = (libgps_gpsdata.status != STATUS_NO_FIX);
+#endif
 		if (gpsdata->valid)
 		{
 			gpsdata->seen_valid = TRUE;
@@ -767,7 +775,11 @@ cb_gpsd_data(GIOChannel *src, GIOCondition condition, 
 			gpsdata->fix.longitude = libgps_gpsdata.fix.longitude;
 			gpsdata->fix.speed = libgps_gpsdata.fix.speed;
 			gpsdata->fix.heading = libgps_gpsdata.fix.track;
+#if GPSD_API_MAJOR_VERSION >= 9
+			gpsdata->fix.altitude = libgps_gpsdata.fix.altMSL;
+#else
 			gpsdata->fix.altitude = libgps_gpsdata.fix.altitude;
+#endif
 		}
 
 		g_source_remove(watchdog);
